src/alertmanager.py

```diff
@@ -0,0 +1,326 @@
+
+    _layer_name = _service_name = _exe_name = "alertmanager"
+
+    # path, inside the workload container for alertmanager data, e.g. 'nflogs', 'silences'.
+    _storage_path = "/alertmanager"
+
+    def __init__(
+        self,
+        charm,
+        *,
+        container_name: str,
+        peer_addresses: List[str],
+        api_port: int,
+        ha_port: int,
+        web_route_prefix: str,
+        external_url: str,
+        config_path: str,
+        web_config_path: str,
+        tls_enabled: bool,
+    ):
+        # Must inherit from ops 'Object' to be able to register events.
+        super().__init__(charm, f"{self.__class__.__name__}-{container_name}")
+
+        self._unit = charm.unit
+
+        self._service_name = self._container_name = container_name
+        self._container = charm.unit.get_container(container_name)
+
+        self._peer_addresses = peer_addresses
+
+        self._api_port = api_port
+        self._ha_port = ha_port
+        self.api = Alertmanager(port=self._api_port, web_route_prefix=web_route_prefix)
+        self._external_url = external_url
+        self._config_path = config_path
+        self._web_config_path = web_config_path
+        self._tls_enabled = tls_enabled
+
+        # turn the container name to a valid Python identifier
+        snake_case_container_name = self._container_name.replace("-", "_")
+        charm.framework.observe(
+            charm.on[snake_case_container_name].pebble_ready,
+            self._on_pebble_ready,
+        )
+
+    @property
+    def is_ready(self):
+        """Is the workload ready to be interacted with?"""
+        return self._container.can_connect()
+
+    def _on_pebble_ready(self, _):
+        if version := self._alertmanager_version:
+            self._unit.set_workload_version(version)
+        else:
+            logger.debug(
+                "Cannot set workload version at this time: could not get Alertmanager version."
+            )
+
+    @property
+    def _alertmanager_version(self) -> Optional[str]:
+        """Returns the version of Alertmanager.
+
+        Returns:
+            A string equal to the Alertmanager version.
+        """
+        if not self.is_ready:
+            return None
+        version_output, _ = self._container.exec([self._exe_name, "--version"]).wait_output()
+        # Output looks like this:
+        # alertmanager, version 0.23.0 (branch: HEAD, ...
+        result = re.search(r"version (\d*\.\d*\.\d*)", version_output)
+        if result is None:
+            return result
+        return result.group(1)
+
+    def check_config(self) -> Tuple[str, str]:
+        """Check config with amtool.
+
+        Returns stdout, stderr.
+        """
+        if not self.is_ready:
+            raise ContainerNotReady(
+                "cannot check config: alertmanager workload container not ready"
+            )
+        proc = self._container.exec(["/usr/bin/amtool", "check-config", self._config_path])
```

Comment:
I would extract `/usr/bin/amtool` to a constant, so it is easy to find in the future.

Reply:
Fixed in next commit.